File: src/QxRibbonWindow.cpp
/**
 * Copyleft (C) 2023 maminjie <canpool@163.com>
 * SPDX-License-Identifier: MIT
**/
#include "QxRibbonWindow.h"
#include "FramelessHelper.h"
#include "QxRibbonBar.h"
#include "QxWindowButtonGroup.h"

#include <QApplication>
#include <QFile>

class RibbonWindowPrivate
{
public:
    RibbonWindowPrivate(RibbonWindow *p);
    void setMenuWidget(QWidget *menuBar);
    void destroyFrameless();
    void setFrameless(bool frameless);
    void resizeRibbon();
public:
    RibbonWindow *q;
    RibbonBar *m_ribbonBar;
    WindowButtonGroup *m_windowButtonGroup;
    FramelessHelper *m_framelessHelper;
    int m_theme;
    bool m_useRibbon;
    bool m_frameless;
};

RibbonWindowPrivate::RibbonWindowPrivate(RibbonWindow *p)
    : q(p)
    , m_ribbonBar(Q_NULLPTR)
    , m_windowButtonGroup(Q_NULLPTR)
    , m_framelessHelper(Q_NULLPTR)
    , m_theme(RibbonTheme::Office2013Theme)
    , m_useRibbon(true)
    , m_frameless(true)
{
}

void RibbonWindowPrivate::setMenuWidget(QWidget *menuBar)
{
    RibbonBar *bar = qobject_cast<RibbonBar *>(menuBar);

    if (bar) {
        m_ribbonBar = bar;
        m_ribbonBar->installEventFilter(q);
        m_useRibbon = true;
    } else {
        m_ribbonBar = Q_NULLPTR;
        m_useRibbon = false;
    }
    setFrameless(m_frameless);
}

void RibbonWindowPrivate::destroyFrameless()
{
    if (m_framelessHelper) {
        delete m_framelessHelper;
        m_framelessHelper = Q_NULLPTR;
    }
    if (m_windowButtonGroup) {
        m_windowButtonGroup->hide();
        delete m_windowButtonGroup;
        m_windowButtonGroup = Q_NULLPTR;
    }
}

void RibbonWindowPrivate::setFrameless(bool frameless)
{
    if (m_useRibbon && frameless) {
        if (Q_NULLPTR == m_framelessHelper) {
            m_framelessHelper = new FramelessHelper(q);
        }
        m_framelessHelper->setTitleHeight(m_ribbonBar->titleBarHeight());
        m_framelessHelper->addCaptionClassName("RibbonBar");
        if (Q_NULLPTR == m_windowButtonGroup) {
            m_windowButtonGroup = new WindowButtonGroup(q);
        }
        Qt::WindowStates s = q->windowState();
        if (s.testFlag(Qt::WindowFullScreen)) {
            m_windowButtonGroup->setVisible(false);
        } else {
            m_windowButtonGroup->setVisible(true);
        }
        m_windowButtonGroup->setWindowStates(s);
        q->setWindowFlags(q->windowFlags() | Qt::FramelessWindowHint);
    } else {
        destroyFrameless();
        q->setWindowFlags(q->windowFlags() & ~Qt::FramelessWindowHint);
    }
}

void RibbonWindowPrivate::resizeRibbon()
{
    if (m_ribbonBar == Q_NULLPTR) {
        return;
    }
    if (m_windowButtonGroup && m_windowButtonGroup->isVisible()) {
        m_ribbonBar->setWindowButtonsSize(m_windowButtonGroup->size());
    } else {
        m_ribbonBar->setWindowButtonsSize(QSize(0, 0));
    }
    m_ribbonBar->resizeRibbon();
}

RibbonWindow::RibbonWindow(QWidget *parent, bool useRibbon)
    : QMainWindow(parent)
    , d(new RibbonWindowPrivate(this))
{
    if (useRibbon) {
        setRibbonTheme(ribbonTheme());
        setMenuWidget(new RibbonBar(this));
        d->resizeRibbon();
    }
}

RibbonWindow::~RibbonWindow()
{
    delete d;
}

RibbonBar *RibbonWindow::ribbonBar() const
{
    return d->m_ribbonBar;
}

void RibbonWindow::setRibbonTheme(int theme)
{
    RibbonTheme::setTheme(theme, this);
    d->m_theme = theme;

    if (d->m_ribbonBar) {
        d->m_ribbonBar->updateRibbonTheme();
    }
}

int RibbonWindow::ribbonTheme() const
{
    return d->m_theme;
}

bool RibbonWindow::isUseRibbon() const
{
    return d->m_useRibbon;
}

bool RibbonWindow::isFrameless() const
{
    return d->m_frameless;
}

void RibbonWindow::setFrameless(bool frameless)
{
    if (d->m_frameless == frameless) {
        return;
    }
    d->m_frameless = frameless;
    d->setFrameless(frameless);
    show(); // the m_windowButtonGroup.isVisible() judgment is added in resizeRibbon(), so show() is called earlier
    d->m_ribbonBar->setWindowTitleVisible(frameless);
    d->resizeRibbon();
}

void RibbonWindow::updateWindowFlags(Qt::WindowFlags flags)
{
    setWindowFlags(flags);
    // Note: This function setWindowFlags calls setParent() when changing the flags for a window,
    // causing the widget to be hidden. You must call show() to make the widget visible again..
    show();
    if (isUseRibbon() && isFrameless()) {
        if (d->m_windowButtonGroup) {
            d->m_windowButtonGroup->updateWindowFlags(flags);
            // FIXME: In the WpsLiteStyle style, there is a short time overlap between rightButtonGroup and
            // windowButtonGroup. Because the button group will be displayed first, and then resize the ribbon.
            // I tried to hide the button group first and display the button group after resize ribbon,
            // but because resize ribbon is a trigger event and non-blocking, the button group was displayed
            // before resize was completed, so I failed.
            d->resizeRibbon();
        }
    }
}

#if QX_RIBBON_DEPRECATED_SINCE(0, 6)
void RibbonWindow::updateWindowFlag(Qt::WindowFlags flags)
{
    updateWindowFlags(flags);
}

Qt::WindowFlags RibbonWindow::windowButtonFlags() const
{
    return windowFlags();
}
#endif

void RibbonWindow::setMenuWidget(QWidget *menuBar)
{
    QMainWindow::setMenuWidget(menuBar);
    d->setMenuWidget(menuBar);
}

void RibbonWindow::setMenuBar(QMenuBar *menuBar)
{
    QMainWindow::setMenuBar(menuBar);
    d->setMenuWidget(menuBar);
}

void RibbonWindow::resizeEvent(QResizeEvent *event)
{
    if (d->m_ribbonBar) {
        if (d->m_ribbonBar->size().width() != this->size().width()) {
            d->m_ribbonBar->setFixedWidth(this->size().width());
        }
        if (d->m_windowButtonGroup && d->m_windowButtonGroup->isVisible()) {
            d->m_ribbonBar->setWindowButtonsSize(d->m_windowButtonGroup->size());
        } else {
            d->m_ribbonBar->setWindowButtonsSize(QSize(0, 0));
        }
    }
    QMainWindow::resizeEvent(event);
}

bool RibbonWindow::eventFilter(QObject *obj, QEvent *e)
{
    if (isFrameless()) {
        // event post ribbonBar -> mainwindow -> framelessHelper
        if (obj == d->m_ribbonBar) {
            switch (e->type()) {
            case QEvent::MouseButtonPress:
            case QEvent::MouseButtonRelease:
            case QEvent::MouseMove:
            case QEvent::Leave:
            case QEvent::HoverMove:
            case QEvent::MouseButtonDblClick:
                QApplication::sendEvent(this, e); // post to framelessHelper
            default:
                break;
            }
        }
    }
    return QMainWindow::eventFilter(obj, e);
}

bool RibbonWindow::event(QEvent *e)
{
    switch (e->type()) {
    case QEvent::WindowStateChange: {
        Qt::WindowStates s = windowState();
        if (isUseRibbon() && isFrameless()) {
            if (d->m_windowButtonGroup) {
                if (s.testFlag(Qt::WindowFullScreen)) {
                    d->m_windowButtonGroup->setVisible(false);
                } else {
                    d->m_windowButtonGroup->setVisible(true);
                }
                d->m_windowButtonGroup->setWindowStates(s);
                d->resizeRibbon();
            }
        }
        emit windowStateChanged(s);
    } break;
    default:
        break;
    }
    return QMainWindow::event(e);
}

void RibbonWindow::loadTheme(const QString &themeFile)
{
    RibbonTheme::loadTheme(themeFile, this);
}

File: pyqt/PyRibbon/QxRibbonWindow.py
#!/usr/bin/env python3
# Copyright (c) 2023 maminjie <canpool@163.com>
# SPDX-License-Identifier: GPL-3.0

from PyQt5.QtCore import QFile, QIODevice
from PyQt5.QtCore import QObject, QEvent
from PyQt5.QtWidgets import QMainWindow, QApplication, QWidget, QMenuBar

from .FramelessHelper import FramelessHelper
from .QxWindowButtonGroup import WindowButtonGroup
from .QxRibbonBar import RibbonBar
from . import QxRibbonRes_rc


class RibbonWindow(QMainWindow):

    def __init__(self, parent=None, use_ribbon=True):
        super().__init__(parent)
        self.m_useRibbon = use_ribbon
        self.m_theme = RibbonWindow.Office2013Theme
        self.m_ribbonBar: RibbonBar = None
        self.m_windowButtonGroup: WindowButtonGroup = None
        self.m_framelessHelper: FramelessHelper = None

        if self.m_useRibbon:
            self.setRibbonTheme(self.ribbonTheme())
            self.setMenuWidget(RibbonBar(self))

    def ribbonBar(self) -> RibbonBar:
        return self.m_ribbonBar

    def setRibbonTheme(self, theme):
        if theme == RibbonWindow.NormalTheme:
            self.loadTheme(':/theme/res/default.qss')
        elif theme == RibbonWindow.Office2013Theme:
            self.loadTheme(':/theme/res/office2013.qss')
        else:
            self.loadTheme(':/theme/res/default.qss')
        self.m_theme = theme

    def ribbonTheme(self) -> int:
        return self.m_theme

    def isUseRibbon(self) -> bool:
        return self.m_useRibbon

    def updateWindowFlag(self, flags):
        if self.isUseRibbon():
            self.m_windowButtonGroup.updateWindowFlag(flags)
        self.repaint()

    def windowButtonFlags(self) -> int:
        if self.isUseRibbon():
            return self.m_windowButtonGroup.windowButtonFlags()
        return self.windowFlags()

    def _setMenuWidget(self, bar: QWidget):
        if isinstance(bar, RibbonBar):
            self.m_ribbonBar: RibbonBar = bar
            self.m_ribbonBar.installEventFilter(self)
            if self.m_framelessHelper is None:
                self.m_framelessHelper = FramelessHelper(self)
            self.m_framelessHelper.setTitleHeight(self.m_ribbonBar.titleBarHeight())
            if self.m_windowButtonGroup is None:
                self.m_windowButtonGroup = WindowButtonGroup(self)
            s = self.m_windowButtonGroup.sizeHint()
            s.setHeight(self.m_ribbonBar.titleBarHeight())
            self.m_windowButtonGroup.setFixedSize(s)
            self.m_windowButtonGroup.setWindowStates(self.windowState())
            self.m_useRibbon = True
            self.m_windowButtonGroup.show()
        else:
            self.m_ribbonBar = None
            self.m_useRibbon = False
            self.m_framelessHelper = None
            if self.m_windowButtonGroup:
                self.m_windowButtonGroup.hide()

    def setMenuWidget(self, bar: QWidget):
        super().setMenuWidget(bar)
        self._setMenuWidget(bar)

    def setMenuBar(self, bar: QMenuBar):
        super().setMenuBar(bar)
        self._setMenuWidget(bar)

    def resizeEvent(self, e):
        if self.m_ribbonBar:
            if self.m_ribbonBar.size().width() != self.size().width():
                self.m_ribbonBar.setFixedWidth(self.size().width())
            if self.m_windowButtonGroup:
                self.m_ribbonBar.setWindowButtonsSize(self.m_windowButtonGroup.size())
        super().resizeEvent(e)

    def eventFilter(self, obj: QObject, e: QEvent) -> bool:
        """
        过滤是为了把 ribbonBar 上的动作传递到 mainwindow，再传递到 frameless
        由于 ribbonbar 会遮挡 frameless 的区域，导致 frameless 无法捕获这些消息
        """
        if obj == self.m_ribbonBar:
            e_list = (
                QEvent.MouseButtonPress,
                QEvent.MouseButtonRelease,
                QEvent.MouseMove,
                QEvent.Leave,
                QEvent.HoverMove,
                QEvent.MouseButtonDblClick
            )
            if e.type() in e_list:
                QApplication.sendEvent(self, e)
        return super().eventFilter(obj, e)

    def event(self, e: QEvent):
        if e and e.type() == QEvent.WindowStateChange:
            if self.isUseRibbon():
                self.m_windowButtonGroup.setWindowStates(self.windowState())
        return super().event(e)

    def loadTheme(self, filepath: str):
        qFile = QFile(filepath)
        if not qFile.open(QIODevice.ReadOnly | QIODevice.Text):
            return
        style_str = str(qFile.readAll(), encoding='utf-8')
        self.setStyleSheet(style_str)
        qFile.close()

    NormalTheme = 0
    Office2013Theme = 1
